src/examples/mnist/MNIST_SP.cpp

```diff
@@ -0,0 +1,272 @@
+/* ---------------------------------------------------------------------
+ * Copyright (C) 2018, David McDougall.
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Affero Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
+ * See the GNU Affero Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero Public License
+ * along with this program.  If not, see http://www.gnu.org/licenses.
+ * ----------------------------------------------------------------------
+ */
+
+/**
+ * Solving the MNIST dataset with Spatial Pooler.
+ * TODO: DESCRIPTION
+ * TODO: RUN INSTRUCTIONS
+ */
+
+#include <algorithm>
+#include <cmath>
+#include <ctime>
+#include <iostream>
+#include <vector>
+#include <random>
+
+#include "nupic/algorithms/SpatialPooler.hpp"
+#include <nupic/algorithms/SDRClassifier.hpp>
+#include <nupic/algorithms/ClassifierResult.hpp>
+#include "nupic/ntypes/SdrMetrics.hpp"
+
+using namespace std;
+using namespace nupic;
+using nupic::algorithms::spatial_pooler::SpatialPooler;
+using nupic::algorithms::sdr_classifier::SDRClassifier;
+using nupic::algorithms::cla_classifier::ClassifierResult;
+
+
+#include <stdio.h>
+#include <execinfo.h>
+#include <signal.h>
+#include <stdlib.h>
+#include <unistd.h>
+void keyboard_interrupt_handler(int sig) {
+  void *array[50];
+  size_t size;
+
+  // get void*'s for all entries on the stack
+  size = backtrace(array, 50);
+
+  // print out all the frames to stderr
+  fprintf(stderr, "Error: signal %d:\n", sig);
+  backtrace_symbols_fd(array, size, STDERR_FILENO);
+  exit(1);
+}
+
+
+vector<UInt> read_mnist_labels(string path) {
```

Comment:
I'd like to rid off this method, and make it an encoder. 
Would you find useful? : 
- MNISTEncoder (little specific, but still useful, hackable for future revisions) 
- Use the mnist dataset in CSV and use our classical CSV parser, as will be used in hotgym